artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java

```diff
@@ -3066,7 +3073,7 @@ private boolean deliver() {
             } else if (!consumers.hasNext()) {
                // Round robin'd all
 
-               if (noDelivery == this.consumers.size()) {
+               if (noDelivery == consumerCount) {
```

Comment:
I had the same worries because my motivation was just to avoid a second related call to size() b/c the logic of seeing noDelivery every where is what I need, it depends on consumers not changing. I can peek some more to get a better answer.
thanks for the shared concern :-)

Reply:
I would leave this with call to size() its a int held number internally of the collection so should have no perf concerns. Is safer than having count being got outside the sync and then consumer added or removed and then its wrong and then theres an introduced concurrency issue

Reply:
leaving the call to size in place